File: 08-Heap-and-Priority-Queue/06-Priority-Queue/src/Array.java
/**
 * 什么是优先队列
 *      普通队列: 先进先出, 后进后出
 *      优先队列: 出队顺序与入队顺序无关, 和优先级相关
 *              类似医院, 根据患者严重程度安排手术
 *              操作系统, 任务的调度, 动态选择优先级最高的任务执行
 *
 *
 * 优先队列
 *                   入队      出队(拿出最大元素)
 *      普通线性结构   O(1)       O(n)
 *      顺序线性结构   O(n)       O(1)
 *      堆           O(logn)    O(logn)
 *
 * 二叉堆是一种满足特殊条件的二叉树
 *      二叉堆是一颗完全二叉树 (完全二叉树: 把元素顺序排列成树的形状)
 *      堆中节点值总是大于等于其左右节点的值(最大堆, 也有最小堆)
 */
public class Array<E> {

    private E[] data;
    private int size;

    public Array(int capacity){
        data = (E[])new Object[capacity];
        size = 0;
    }

    public Array(){
        this(10);
    }

    public Array(E[] arr){
        data = (E[])new Object[arr.length];
        for(int i = 0; i < arr.length; i++)
            data[i] = arr[i];
        size = arr.length;
    }

    public int getCapacity(){
        return data.length;
    }

    public int getSize(){
        return size;
    }

    public boolean isEmpty(){
        return size == 0;
    }

    public void add(int index, E e){
        if (index < 0 || index > size)
            throw new IllegalArgumentException("Add failed. index Illegal");

        if (size == data.length)
            resize(2*data.length);

        for (int i=size-1; i>=index; i--)
            data[i+1] = data[i];

        data[index] = e;
        size++;
    }

    public void addLast(E e){
        add(size,e);
    }

    public void addFirst(E e){
        add(0,e);
    }

    public E get(int index){
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("Get failed. index Illegal");

        return data[index];
    }

    public void set(int index, E e){
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("Set failed. index Illegal");
        data[index] = e;
    }

    public E remove(int index){
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("Remove failed. index Illegal");

        E ret = data[index];

        for (int i=index+1; i<size; i++)
            data[i-1] = data[i];

        size--;
        data[size] = null;
        if (size == data.length/4 && data.length / 2 != 0)
            resize(data.length/2);

        return ret;
    }

    public E removeLast(){
        return remove(size-1);
    }

    private void resize(int newCapacity){

        E[] newData = (E[])new Object[newCapacity];
        for(int i=0; i<size; i++)
            newData[i] = data[i];
        data = newData;
    }

    public void swap(int i, int j){
        if (i < 0 || i >= size || j < 0 || j >= size)
            throw new IllegalArgumentException("index Illegal");

        E t = data[i];
        data[i] = data[j];
        data[j] = t;
    }
}

File: go/02-Arrays/09-Amortized-Time-Complexity/array/array.go
package array

import (
	"bytes"
	"fmt"
	"github.com/dingsw2019/play-with-data-struct/go/utils"
)

type Array struct {
	data []interface{}
	size int
}

func New(capacity int) *Array {
	return &Array{
		data: make([]interface{}, capacity),
	}
}

// 获取容量
func (a *Array) GetCapacity() int {
	return len(a.data)
}

// 获取元素数量
func (a *Array) GetSize() int {
	return a.size
}

// 是否空数组
func (a *Array) IsEmpty() bool {
	return a.size == 0
}

// 任意位置添加元素
func (a *Array) Add(index int, e interface{}) {
	if index < 0 || index > a.size {
		panic("add failed, index out of range")
	}
	// 扩容
	if a.size == len(a.data) {
		a.resize(2 * a.size)
	}
	// 移动元素
	for i := a.size - 1; i >= index; i-- {
		a.data[i+1] = a.data[i]
	}
	a.data[index] = e
	a.size++
}

// 尾部添加元素
func (a *Array) AddLast(e interface{}) {
	a.Add(a.size, e)
}

// 头部添加元素
func (a *Array) AddFirst(e interface{}) {
	a.Add(0, e)
}

// 获取元素
func (a *Array) Get(index int) interface{} {
	a.checkIndex(index)
	return a.data[index]
}

// 更改元素值
func (a *Array) Set(index int, e interface{}) {
	a.checkIndex(index)
	a.data[index] = e
}

// 是否包含元素
func (a *Array) Contains(e interface{}) bool {
	for i := 0; i < a.size; i++ {
		if utils.Compare(a.data[i], e) == 0 {
			return true
		}
	}
	return false
}

// 查找元素所在位置
func (a *Array) Find(e interface{}) int {
	for i := 0; i < a.size; i++ {
		if utils.Compare(a.data[i], e) == 0 {
			return i
		}
	}
	return -1
}

// 查找相同元素的所有索引位
func (a *Array) FindAll(e interface{}) (indexes []int) {
	for i := 0; i < a.size; i++ {
		if utils.Compare(a.data[i], e) == 0 {
			indexes = append(indexes, i)
		}
	}
	return
}

// 删除元素
func (a *Array) Remove(index int) interface{} {
	a.checkIndex(index)
	e := a.data[index]
	for i := index; i < a.size; i++ {
		a.data[i] = a.data[i+1]
	}
	a.size--

	// 缩容, 1/4保证不会在临界值重复扩容缩容
	// size == 1 时, 保证resize不会为0
	if a.size == len(a.data)/4 && len(a.data)/2 != 0 {
		a.resize(len(a.data) / 2)
	}
	return e
}

// 头部删除
func (a *Array) RemoveFirst() interface{} {
	return a.Remove(0)
}

// 尾部删除
func (a *Array) RemoveLast() interface{} {
	return a.Remove(a.size - 1)
}

// 删除指定元素
func (a *Array) RemoveElement(e interface{}) bool {
	index := a.Find(e)
	if index == -1 {
		return false
	}
	a.Remove(index)
	return true
}

// 删除所有相同的指定元素
func (a *Array) RemoveAllElement(e interface{}) bool {
	start := a.Find(e)
	if start == -1 {
		return false
	}
	for i := start; i < a.size; i++ {
		if utils.Compare(a.data[i], e) == 0 {
			a.Remove(i)
		}
	}
	return true
}

// 重设容量
func (a *Array) resize(newCapacity int) {
	newData := make([]interface{}, newCapacity)
	for i := 0; i < a.size; i++ {
		newData[i] = a.data[i]
	}
	a.data = newData
}

// 校验索引值
func (a *Array) checkIndex(index int) {
	if index < 0 || index >= a.size {
		panic(fmt.Sprintf("index out of range, required range [0,%d], but get %d", a.size, index))
	}
}

// 重写String
func (a *Array) String() string {
	var buf bytes.Buffer

	buf.WriteString(fmt.Sprintf("Array: size = %d, capacity = %d\n", a.size, len(a.data)))
	buf.WriteString("[")
	for i := 0; i < a.size; i++ {
		buf.WriteString(fmt.Sprintf("%v", a.data[i]))
		if i != a.size-1 {
			buf.WriteString(", ")
		}
	}
	buf.WriteString("]")
	return buf.String()
}
